src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+                // See following link:
+                // https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposable-transient-services-captured-by-container
+                throw new InvalidOperationException("A schema that implements IDisposable cannot be registered as a transient service.");
+            }
+
+            // Register the service with the DI provider as TSchema, overwriting any existing registration
+            builder.Register(schemaFactory, serviceLifetime);
+
+            // Now register the service as ISchema if not already registered.
+            builder.TryRegister<ISchema>(schemaFactory, serviceLifetime);
+
+            return builder;
+        }
+        #endregion
+
+        #region - AddDocumentExecuter -
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter, TDocumentExecuter>(ServiceLifetime.Singleton);
+
+        /// <summary>
+        /// Registers <paramref name="documentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder, TDocumentExecuter documentExecuter)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter>(documentExecuter ?? throw new ArgumentNullException(nameof(documentExecuter)));
+
+        /// <summary>
+        /// Registers <typeparamref name="TDocumentExecuter"/> as a singleton of type <see cref="IDocumentExecuter"/> within the
+        /// dependency injection framework. The supplied factory method is used to create the document executer.
+        /// </summary>
+        public static IGraphQLBuilder AddDocumentExecuter<TDocumentExecuter>(this IGraphQLBuilder builder, Func<IServiceProvider, TDocumentExecuter> documentExecuterFactory)
+            where TDocumentExecuter : class, IDocumentExecuter
+            => builder.Register<IDocumentExecuter>(documentExecuterFactory ?? throw new ArgumentNullException(nameof(documentExecuterFactory)), ServiceLifetime.Singleton);
+        #endregion
+
+        #region - AddComplexityAnalyzer -
+        /// <summary>
+        /// Enables the default complexity analyzer and configures it with the specified configuration delegate.
+        /// </summary>
+        public static IGraphQLBuilder AddComplexityAnalyzer(this IGraphQLBuilder builder, Action<ComplexityConfiguration>? action = null)
```

Comment:
Something wrong here. I don't see that analyzer is added. Also I don't see that complexity configuration is modified anywhere. The same for the next method and "for half" for third.

Reply:
OK. I found that `IComplexityAnalyzer` is registered in GraphQLBuilder ctor as _default_ service. But the question about modification of complexity configuration remains.

Reply:
Correct, the default complexity analyzer is registered as a default service.  So the standard AddComplexityAnalyzer does not overwrite the registration in case you have already registered another complexity analyzer.

I considered using `Configure<T>` to configure options and setting them in the execution.  However, this will always create an empty instance and then call the configuration delegates.  If the MS Options abstractions were available this perhaps could be solved, but without them, we are faced with either (a) ignoring the current `options.ComplexityConfiguration` settings and creating a new instance, or (b) ignoring the options abstractions and invoking the delegate directly.  So that's what I did:

```cs
public static IGraphQLBuilder AddComplexityAnalyzer(this IGraphQLBuilder builder, Action<ComplexityConfiguration>? action = null)
    => builder.ConfigureExecution(opts =>
    {
        opts.ComplexityConfiguration ??= new ComplexityConfiguration();
        action?.Invoke(opts.ComplexityConfiguration);
    });
```

You see above that all the method does is register a delegate when ExecuteAsync runs, and passes the complexity configuration class to the `action` delegate to be mutated as desired.  This preserves any complexity settings the user may have already had when calling ExecuteAsync.

For v5, I suggest we remove ComplexityConfiguration from ExecutionOptions and instead have the options fed via DI to a new instance of the complexity analyzer, much like ErrorInfoProvider or SystemTextJson, etc.

We might also, for v5, consider removing all complexity analyzer code entirely -- and instead putting the complexity analyzer into a validation rule.  The builder method can remain as it is here, but it will then just be calling `AddValidationRule` along with `Configure` to configure the options.  This has a benefit of simplifying the document executer, and allowing the complexity analyzer to skip execution if the document is pulled from the document cache.

It is noteworthy that if a person uses the v4 builder methods to configure the complexity analyzer, then no changes should be required for v5 however we refactor the complexity analyzer.  (Hopefully.)

Reply:
#2791